Storage: Add Oracle Cloud Infrastructure Object Storage Bucket support

aarontams opened the pull request:
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes
#### Add Oracle Cloud Infrastructure (OCI) Object Storage Bucket support
OCI Object Storage supports three different providers, default, instance principle, and raw. A lot of OCI users like to use Instance principle to security. It is very important to support instance principle for the OCI users. `S3 doesn't support OCI instance principle.` 

#### Support multiple prometheus instances with a single bucket 
By default, like other clients, this new Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. This means each bucket can only store one prometheus instance data.

In OCI, a lot of users like to put different type of data in a single bucket.
To support multiple prometheus instances with a single bucket, a new object_base_path option is introduced in this OCI Object Storage client. If defined, the value will be used as the prefix of the prometheus data in OCI bucket.

For example, if you want to use the same bucket for 2 prometheus instances, you can define object_base_path as /path/for/prometheus1 for the first prometheus instance, and /path/for/prometheus2 for the second prometheus instance.

Three tests are included here to demo how the object_base_path work.

## Verification
Running similar OCI client against 3 prometheus instances for over 50 days without any issue.
For each prometheus instance
- average 4.5GB (2GB in chunks and 2.5G index) per 2-hour prometheus block
- over 50GB per day
- over 2TB prometheus data uploaded to OCI Object Storage
<img width="520" alt="oci storage 1" src="https://user-images.githubusercontent.com/20829640/133557741-4a85e6e5-7539-43e2-93fc-320d8e252e1d.png">
<img width="917" alt="oci storage 2" src="https://user-images.githubusercontent.com/20829640/133557759-34ea339e-ddd1-4a68-977a-ff40d124c2d5.png">

### Testing
```
export THANOS_TEST_OBJSTORE_SKIP=GCS,S3,AZURE,SWIFT,COS,ALIYUNOSS,BOS
go test -v -run TestObjStore_AcceptanceTest_e2e ./pkg/...
```
See the test steps in storage.md under OCI section.

Ran three tests with different OCI_OBJECT_BASE_PATH settings. Below are the screenshots from Oracle Cloud Infrastructure console to show the object paths when the tests were running.
 
#### Test 1 - OCI_OBJECT_BASE_PATH is not set
<img width="947" alt="test1" src="https://user-images.githubusercontent.com/20829640/133550627-5218287b-cb6e-45b5-8ebd-44646903e416.png">

#### Test 2 - OCI_OBJECT_BASE_PATH=/my/base/path
<img width="958" alt="test2" src="https://user-images.githubusercontent.com/20829640/133550651-4e650fbb-65fb-4fda-80a5-837c66647d94.png">

#### Test 3 - OCI_OBJECT_BASE_PATH=test3/base-path
 <img width="954" alt="test3" src="https://user-images.githubusercontent.com/20829640/133550697-b8c0d746-1ac0-42c5-ade1-22967e84d9bb.png">

aarontams commented:
Here is `docs / Documentation check` error:
```
error: 		docs/storage.md:423: "https://cloud.baidu.com/doc/BOS/index.html" not accessible; status code 0: Get "https://cloud.baidu.com/doc/BOS/index.html": context deadline exceeded
```

The `https://cloud.baidu.com/doc/BOS/index.html ` link is not belong to my MR. The error is from the original codes.

aarontams commented:
**For ci/circleci testing**
To test OCI, you will need to provide real OCI Objectstorage bucket, otherwise the test will fail.

aarontams commented:
**For go / Linters (Static Analysis) for Go**
```
cannot 'detected not clean work tree before running lint: you have unstaged changes.
M	pkg/objstore/oci/helper.go
M	pkg/objstore/oci/oci.go
Please commit or stash them.
```
I assume this is normal for a PR. Please let me know if this is not the case here. Thx!

aarontams commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,331 @@
+//NewTestBucket creates test bkt client that before returning creates temporary bucket.
+//In a close function it empties and deletes the bucket.
+func NewTestBucket(t testing.TB) (objstore.Bucket, func(), error) {
+	config, err := getConfigFromEnv()
```
Add error checking
```
	config, err := getConfigFromEnv()
	if err != nil {
		return nil, nil, err
	}
```

GiedriusS commented:
Please add the DCO (:

aarontams commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,332 @@
+		ResponseHeaderTimeout: time.Duration(config.HTTPConfig.ResponseHeaderTimeout),
+		TLSHandshakeTimeout:   time.Duration(config.HTTPConfig.TLSHandshakeTimeout),
+		ExpectContinueTimeout: time.Duration(config.HTTPConfig.ExpectContinueTimeout),
+		TLSClientConfig:       &tls.Config{InsecureSkipVerify: config.HTTPConfig.InsecureSkipVerify},
```
ignore

aarontams commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,331 @@
+}
+
+// createBucket creates bucket.
+func (b *Bucket) createBucket(ctx context.Context, compartmentId string) (err error) {
```
ignore 
Use by test.

aarontams commented:
Please let me know if there is any additional information that I need to provide to help the review process. Thx!

wiardvanrij commented:
Hi, looks great and I think it makes sense to provide wider support for different clouds.

> By default, like other clients, this new Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. This means each bucket can only store one prometheus instance data. 

Is it though? I'm quite confident that a specific ULID is created and it should not be a problem to use the same object store for n-amount of instances. If one would like to separate the data, one could use a different object store instance. 

I'm not really a fan of the `object_base_path`.  It's prone for (user) error; what if someone does NOT change the value of it, what happens then? 

Perhaps you could elaborate a bit more on it for me, it might be that I'm missing the actual point of course :)

aarontams commented:
> Hi, looks great and I think it makes sense to provide wider support for different clouds.
> 
> > By default, like other clients, this new Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. This means each bucket can only store one prometheus instance data.
> 
> Is it though? I'm quite confident that a specific ULID is created and it should not be a problem to use the same object store for n-amount of instances. If one would like to separate the data, one could use a different object store instance.
> 
> I'm not really a fan of the `object_base_path`. It's prone for (user) error; what if someone does NOT change the value of it, what happens then?
> 
> Perhaps you could elaborate a bit more on it for me, it might be that I'm missing the actual point of course :)

_**To answer your initial question.**_ 
`object_base_path` is an optional field. If it is not provided, this new client will act the same as all other existing storage clients. So, if you don't want it, you can completely ignore it in your store setup.

_**Here is use case for our team:**_
In our production env, we have over 40 K8s clusters. Some of the K8s cluster have over 100 prometheus instances. For each K8s cluster, we create one bucket in OCI. With the `object_base_path`, we can potentially capture data from all 100 prometheus instances in a single bucket.

We use OCI instance principle to grant bucket access to the K8s cluster. We also create/delete prometheus instances dynamically (on demand) all the time. With the `object_base_path`, we can save some time/work to set up or delete OCI buckets when the number of prometheus instances changed.

In additional, with the `object_base_path`, we can easily find and delete all the data for a specific prometheus. For example:
- prometheus_001
- prometheus_002
- ....
- prometheus_123

I am open for suggestion for my MR. If you really think `object_base_path` shouldn't be part of the offer even it is just an optional field, I will consider to remove it (or hide it from the configuration) so I can have the basic implementation live in the Thanos github repo. In this case, I just need to add the little enhancement each time I upgrade Thanos in our production env. Please let me know what you think.

I have the similar code running in few of my prometheus instances for over couple months with hiccup 🤞 

Again, thanks for reviewing my MR.

aarontams commented:
Thanks. Looking forward to seeing more reviewers.

If it helps, I am happy to do a quick demo during the Thursday US community meeting in couple weeks so reviewers/people can better understand the motivation of this MR and how things work in our team and OCI. Just let me know here and feel free to slack me in #thanos-dev.

aarontams commented:
Hi @wiardvanrij  @GiedriusS @bwplotka 
I tried to attend the Thanos community meeting yesterday to get some help for this PR, but the meeting got cancelled. Please let me know if there is anything that I can do to move this PR forward. 
Thanks in advance.

GiedriusS commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,335 @@
+			return nil, errors.Wrapf(err, "unable to create OCI instance principal config provider")
+		}
+	case RawConfigProvider:
+		if err := config.validateConfig(); err != nil {
```
Do we need to call this in all cases? If not, do we need extra validation for other providers?

aarontams replied:
InstancePrincipalConfigProvider 
The OCI Instance Principe setup is when you setup the OCI objectstorage bucket. You grant the permission through OCI policies. All the checking with by covered by auth.InstancePrincipalConfigurationProvider. A nice error message will be thrown if the instance principal policy configuration has issue. There  is no need to duplicate the same logic that OCI GO SDK already covered.
See https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm.

DefaultConfigProvider
This is the most standard way to connect to OCI especially for users run stuff from their Mac (local marine). Typically, it uses the local .oci/config file to configuration the OCI provider. The common.DefaultConfigProvider() also provided a very nice error message. There  is no need to duplicate the same logic that OCI GO SDK already covered.

RawConfigProvider
Most of the OCI user don't use provider because this API requires you to pass few parameters (such as Fingerprint). This provider is for the user who want to create the OCI provider on the fly which is the vey common (and secure) in OCI world. My `validateConfig()` only covers the most basic one to remind the users what are the minimum requirements. My thinking it would be nice to spell it out because as you can see the `common.NewRawConfigurationProvider()` call requires some parameters (unlike `common.DefaultConfigProvider()` and `auth.InstancePrincipalConfigurationProvider()` which do not require any parameter). I know OCI GO SDK does provide some good error message.

If you want to make all 3 provider looks alike, I think I just remove the config.validateConfig() completely.
```
if err := config.validateConfig(); err != nil {
			return nil, errors.Wrapf(err, "invalid oci configurations")
		}
```

aarontams replied:
Please let me know if you still have concern with this implementation.

GiedriusS replied:
Thank you for this really thorough explanation, looks good!

aarontams commented:
Thanks @GiedriusS for the review.
I will update my MR and provide more info for your comments when I have a chance.

aarontams commented on docs/storage.md:
```diff
@@ -449,6 +450,112 @@ config:
+OCI Object Storage supports three different providers, default, instance principle, and raw. 
+The default HTTP Client comes with this Thanos OCI Object Storage should be able to handle most of the use cases.
+
+By default, like other clients, Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. 
```
I will fix it in my next push (I may wait for your feedback from other comments first).

aarontams replied:
Pushed the fix.

aarontams commented on docs/storage.md:
```diff
@@ -449,6 +450,112 @@ config:
+The default HTTP Client comes with this Thanos OCI Object Storage should be able to handle most of the use cases.
+
+By default, like other clients, Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. 
+This means each bucket can only store one prometheus instance data.
```
Sorry for the poor description.
When thanos uploads all the blocks from a prometheus instance, it will create a directory structure similar to the following in OCI objectstore
```text
|-- 01FF8QX35297NG9HCDHGH1BN1
|   |-- chunks
|       |-- 000001
|       |-- 000002
|       |-- ....
|-- 01FFPFKMAQQNF4N945X7JWPQ8
|   |-- chunks
|.      |-- ....
|-- .... other prometheus blocks
|
|-- debug
    |-- meta
        |-- .....
```
`I consider above as one prometheus instance data.`

In this PR, a new option called object_base_path is added.
If defined, the whole directory structure above will be uploaded under the object_base_path prefix.

In second screenshot in this PR, it holds 4 prometheus instances.
Prometheus A - object_base_path is set to thanos.dev.thanos
Prometheus B - object_base_path is set to thanos.meta.sa/0
Prometheus C - object_base_path is set to thanos.meta.sa/1
Prometheus D - object_base_path is set to thanos.meta.sa/2

Here is the illustration in OCI objectstorage bucket.
```
|-- thanos.dev.thanos
    |--  <<Promethus-A with all blocks and debug meta>>
|-- thanos.meta.sa
    |-- 0 <<Promethus-B with all blocks and debug meta>>
    |-- 1 <<Promethus-C with all blocks and debug meta>>
    |-- 2 <<Promethus-D with all blocks and debug meta>>
        |-- 01FF8QX35297NG9HCDHGH1BN1
        |   |-- chunks
        |       |-- 000001
        |       |-- 000002
        |       |-- ....
        |-- 01FFPFKMAQQNF4N945X7JWPQ8
        |   |-- chunks
        |.      |-- ....
        |-- .... other prometheus blocks
        |
        |-- debug
            |-- meta
                |-- .....
```

If you want to provide a better description for the Thanos community, I will be happy to change it in my PR.

GiedriusS replied:
Yes, that's what I mean - this https://github.com/thanos-io/thanos/issues/1318 has been a long-standing issue. I think this kind of functionality needs to be added "one level above" i.e. not at the raw object storage calls level. Otherwise, it would be weird to have duplicate functionality. I think others would agree as well. My suggestion would be to remove this prefix functionality and move it one level above so that it could be used with any kind of object storage. It could be done in a separate pull request. We could double check in #thanos-dev, though

aarontams replied:
Got it. Yes, adding the feature to `one level up` is a great idea for all Objectstorage clients.

Since allowing the same bucket to hold multiple prometheus instances' data is the main use case of our team and we would like to use Thanos master branch instead of my forked branch.

I can think of three different approaches (order by my preferences) to avoid duplicate functionality in my PR:
**_Approach # 1_**
- hide the object_base_path configuration from storage.md documentation
- without the object_base_path setting, my PR will act exactly the same as other ObjecctStorage Clients
- the hidden feature can still be used by the insider who know the existence of this feature
- after the code check it, whoever want to work on #1318 will have an example how OCI client is implemented.

**_Approach # 2_**
- hide the object_base_path configuration from storage.md documentation
- hard code the object_base_path to empty string and won't accept any hidden customization

**_Approach # 3_**
Remove all the object_base_path throughout the PR.


Please let me know which approach is acceptable for you.
Thanks!

wiardvanrij replied:
Sorry for the delay again, bit busy times. 
As I like the different approaches I'm not that fan of hidden features in a way that one needs to know the existence in an open source project. Also I took some time to read the long standing issue and the consensus after the discussion was this:

> Looks like the decision is that we are ok for simple custom prefix on config level (for all objstorages) in YAML (:

So, I would say that if we implement it, it should be "that". Still open for discussion though, but then it should require a lot more discussion from everyone else again IMO ^^

I would say, let's just implement this on level higher as discussed in that main issue. For all the storage backends, implement at least the config in a way that it can say "This is not supported yet for this object store" if one would configure it. With then the exception for OCI, in which you did implement it. 

This gives a bit more motivation to get that issue fixed, and we can have iterations per objectstorage to actually implement it. I think by saying "It needs to be done for them all in one go", is really limiting us to get it moving forward. 

What do you think? Also, I'm actually happy to help on that. I got some spare time soon.

GiedriusS replied:
I really think we should do (3) approach here. Then, you could implement it in a follow up PR. We have interfaces for object storage so IMHO it should be fairly easy to implement prefix functionality.

aarontams replied:
I will remove the prefix option when I have a chance.

GiedriusS replied:
Don't forget to remove this sentence as it is not true in practice, it is _very_ hard to get a UUID collision.

aarontams commented on docs/storage.md:
```diff
@@ -449,6 +450,112 @@ config:
+
+#### Default Provider
+The default config provider will look for configurations in 3 places: file in `$HOME/.oci/config`, `$HOME/.obmcs/config`, and
+variables names starting with the string TF_VAR. If the same configuration is found in multiple places the provider will
```
Yes, TF_VAR is referring to the OCI environment variables that are used to connect to OCI.
For users who use OCI, they should be very familiar with the name TF_VAR. Probably, TF_VAR will be the first thing that they learn when using OCI outside of OCI Console.

This is the verbatim from the public OCI GO SDK APIs. For this PR, I tried to be consistent with the definition in OCI GO SDK so the users could feel like home.

See https://github.com/oracle/oci-go-sdk/blob/master/common/client.go#L301 for details.

GiedriusS replied:
:+1:

aarontams commented on docs/storage.md:
```diff
@@ -449,6 +450,112 @@ config:
+#### Default Provider
+The default config provider will look for configurations in 3 places: file in `$HOME/.oci/config`, `$HOME/.obmcs/config`, and
+variables names starting with the string TF_VAR. If the same configuration is found in multiple places the provider will
+prefer the first one.
```
See reply from above for full `Default Provider` explanation.

aarontams commented on docs/storage.md:
```diff
@@ -449,6 +450,112 @@ config:
+  provider: "default"
+  bucket: ""
+  compartment_ocid: ""
+  object_base_path: ""            // Optional object base path (prefix) in OCI bucket.
```
Please see the explanation above for  _**What does this mean? Why only one?**_
Hope it will make more sense after you read my explanation.

If you want to use different variable name or an alternative approach for this new objectstorage prefix feature, please let me know.

aarontams commented on docs/storage.md:
````diff
@@ -449,6 +450,112 @@ config:
+```
+You can also include any of the optional configuration just like the example in `Default Provider`.
+
+#### TestObjStore_AcceptanceTest_e2e
````
I will remove this section in my next push (I will wait for your feedback from other comments first).

aarontams replied:
Pushed the fix.

aarontams commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,332 @@
+
+func getNamespace(client objectstorage.ObjectStorageClient, requestMetadata common.RequestMetadata) (namespace *string, err error) {
+	response, err := client.GetNamespace(
+		context.Background(),
```
In OCI GO SDK, the ObjectStorage Client already has a nice build it timeout and retry mechanism. It will be kind of redundant if we do it here. It was suggested by the internal OCI SDK team.
See https://github.com/oracle/oci-go-sdk/blob/5d8aab248378a0fcf7cab02926b696e6c8d34811/objectstorage/objectstorage_client.go#L983.

GiedriusS replied:
AFAICT under the hood it uses the same `http.Transport` that we pass to it in the constructor and it has no timeouts? :thinking: Could you please show me where the timeouts are set for this operation? I don't think there is. But, I think we should stick with this `context.Background` because this is just a one-time operation on boot.

aarontams replied:
httpClient is created by calling `DefaultTransport(config Config)`
```
	httpClient := http.Client{
		Transport: DefaultTransport(config),
	}
```

The DefaultTransport() below has `IdleConnTimeout`, `ResponseHeaderTimeout`, `TLSHandshakeTimeout`, and `ExpectContinueTimeout`.
```
func DefaultTransport(config Config) *http.Transport {
	return &http.Transport{
		Proxy: http.ProxyFromEnvironment,
		DialContext: (&net.Dialer{
			Timeout:   30 * time.Second,
			KeepAlive: 30 * time.Second,
			DualStack: true,
		}).DialContext,

		IdleConnTimeout:       time.Duration(config.HTTPConfig.IdleConnTimeout),
		ResponseHeaderTimeout: time.Duration(config.HTTPConfig.ResponseHeaderTimeout),
		TLSHandshakeTimeout:   time.Duration(config.HTTPConfig.TLSHandshakeTimeout),
		ExpectContinueTimeout: time.Duration(config.HTTPConfig.ExpectContinueTimeout),
		TLSClientConfig:       &tls.Config{InsecureSkipVerify: config.HTTPConfig.InsecureSkipVerify},
		MaxIdleConns:          config.HTTPConfig.MaxIdleConns,
		MaxIdleConnsPerHost:   config.HTTPConfig.MaxIdleConnsPerHost,
		MaxConnsPerHost:       config.HTTPConfig.MaxConnsPerHost,
		DisableCompression:    config.HTTPConfig.DisableCompression,
	}
}
```

Is that what you are looking for?

GiedriusS replied:
No, I am talking about the timeout of this operation. The client has no timeout and we pass `context.Background` here so in case of a very slow TCP read, for example, this will be stuck indefinitely, right? :thinking: Perhaps we could add a log message before calling this function?

GiedriusS replied:
Still missing a log message here before this call, with the `debug` log level.

GiedriusS commented:
@aarontams sorry for the delay, there's a lack of reviewers/time. I'll try to get back soon.

aarontams commented:
Thanks @GiedriusS 

I just updated my MR 
- rebase to thanos main/master
- upgrade oci-go-sdk to v54.0.0
- add retry option in case of hitting OCI rate limit

I also chatted with @wiardvanrij before the holiday regarding may be having the `objectBasePath` framework for all provider. We said we may look into it together this year.

GiedriusS commented:
AFAICT unit test failures are related to your code. Could you please take a look?

GiedriusS commented:
CI failure seems related. Could you please fix the formatting with `make format`? I also think we should remove the prefix functionality from this. Because we have the objstore interface, it should be fairly easy to implement it in one level above this.

aarontams reviewed (commented):
I hope I have addressed all the comments and concerns.

When I have a chance, I will remove the `object_base_path` option.

Djelibeybi commented on docs/storage.md:
```diff
@@ -468,6 +469,81 @@ config:
+You can get a free OCI Cloud Services by following the [Oracle Cloud Free Tier](https://www.oracle.com/cloud/free) page.
+More information about [Overview of Object Storage](https://docs.cloud.oracle.com/en-us/iaas/Content/Object/Concepts/objectstorageoverview.htm).
+
+OCI Object Storage supports three different providers, default, instance principle, and raw. 
```
This should be instance _principal_.

aarontams replied:
Thanks for the eagle eyes. Push the fix a minute ago.

Djelibeybi commented on docs/storage.md:
````diff
@@ -468,6 +469,81 @@ config:
+```
+
+
+#### Instance Principle Provider
````
This should be _Principal_ as per the `provider` value below. :)

aarontams replied:
Same, pushed the fix a minute ago.

aarontams commented:
I believe my last push should fix both CI and format issue.
Note: There is a potential known issue with the OCI Instance principle. I added a retry f

I will remove the `prefix functionality` from the new OCI Objectstorage API.

aarontams commented:
@GiedriusS and @wiardvanrij 
Per @GiedriusS 's suggestion
>  I also think we should remove the prefix functionality from this. Because we have the objstore interface, it should be fairly easy to implement it in one level above this.

I removed `object_base_path` from my PR.  Hopefully, we will have the same implementation in one level up soon. Thx!

GiedriusS commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,331 @@
+}
+
+func getObject(ctx context.Context, bkt Bucket, objectName string, byteRange string) (response objectstorage.GetObjectResponse, err error) {
+	// path should not be empty
```
Nit: let's remove this, we can see what the `if` is doing

aarontams replied:
Removed.

GiedriusS commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,331 @@
+	}
+
+	provider := Provider(strings.ToLower(config.Provider))
+	level.Info(logger).Log("msg", fmt.Sprintf("Creating OCI default '%s' provider", provider))
```
Nit: the pattern is to specify separate fields as separate arguments so it should be:
```suggestion
	level.Info(logger).Log("msg", "creating OCI client", "provider", provider)
```

aarontams replied:
Changed.

GiedriusS commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,331 @@
+		partSize:        config.PartSize,
+		requestMetadata: requestMetadata,
+	}
+	level.Info(logger).Log("msg", fmt.Sprintf("[Bucket] name: '%s'", bkt.name))
```
Do we need to log this? (:

aarontams replied:
Not really. Just my bad habit :)

Removed.

GiedriusS commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,331 @@
+	client.HTTPClient = &httpClient
+
+	requestMetadata := getRequestMetadata(config.MaxRequestRetries, config.RequestRetryInterval)
+
```
Let's add here:

```go
level.Info(logger).Log("msg", "getting namespace, it might take some time")
```

aarontams replied:
Added.

GiedriusS reviewed (commented):
LGTM modulo some nits regarding logs / comments. Let's remove the prefix functionality and remove prefix stuff from the docs. Plus, linter is not lying (: then we can merge this!

aarontams commented:
@GiedriusS ,
Thanks for all the comments, I believe I have addressed them in my new push few minutes ago.

I also fix the go-lint warnings in my new push except the following
```
Error: pkg/objstore/oci/oci.go:232:18: func `(*Bucket).deleteBucket` is unused (unused)
413
func (b *Bucket) deleteBucket(ctx context.Context) (err error) {
414
                 ^
415
Error: pkg/objstore/oci/oci.go:227:18: func `(*Bucket).createBucket` is unused (unused)
416
func (b *Bucket) createBucket(ctx context.Context, compartmentId string) (err error) {
```
This two APIs are used by the test and I don't think I can removed them.

GiedriusS commented:
> @GiedriusS , Thanks for all the comments, I believe I have addressed them in my new push few minutes ago.
> 
> I also fixed the go-lint warnings in my new push except the following
> 
> ```
> Error: pkg/objstore/oci/oci.go:232:18: func `(*Bucket).deleteBucket` is unused (unused)
> 413
> func (b *Bucket) deleteBucket(ctx context.Context) (err error) {
> 414
>                  ^
> 415
> Error: pkg/objstore/oci/oci.go:227:18: func `(*Bucket).createBucket` is unused (unused)
> 416
> func (b *Bucket) createBucket(ctx context.Context, compartmentId string) (err error) {
> ```
> 
> This two APIs are used by the test and I don't think I can removed them.

Test is calling `deleteBucket()` directly. `Bucket.deleteBucket` is not called. To fix this, you'll need to change that into `bkt.deleteBucket(ctx)`.

GiedriusS commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,330 @@
+	t.Logf("created temporary OCI bucket '%s' for OCI tests", bkt.name)
+	return bkt, func() {
+		objstore.EmptyBucket(t, ctx, bkt)
+		if err := deleteBucket(ctx, bkt); err != nil {
```
```suggestion
		if err := bkt.deleteBucket(ctx); err != nil {
```

aarontams replied:
changed.

GiedriusS commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,330 @@
+
+	ctx := context.Background()
+	bkt.name = objstore.CreateTemporaryTestBucketName(t)
+	if err := createBucket(ctx, bkt, config.Compartment); err != nil {
```
To fix the linter:
```suggestion
	if err := bkt.createBucket(ctx, config.Compartment); err != nil {
```

aarontams replied:
changed.

GiedriusS commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,327 @@
+	return
+}
+
+func deleteBucket(ctx context.Context, bkt *Bucket) (err error) {
```
In fact, why do we separate functions for this? Just move this into `Bucket.deleteBucket()` and that's all (:

aarontams replied:
Same reply as above

aarontams commented:
@GiedriusS 
Thanks for all the comments. I will address them this weekend.

aarontams commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,327 @@
+	return
+}
+
+func createBucket(ctx context.Context, bkt *Bucket, compartmentId string) (err error) {
```
There are two go files in my PR - oci.go and helper.go.

All the Bucket interface APIs are in oci.go. To keep it clean and organize, I put the all the OCI SDK calls (requests) in helper.go.

My thinking is people can see all the OCI SDK calls in a single file instead all over the places.  But if you still think this is confusing, I am fine with moving the createBucket() and deleteBucket() codes to oci.go. Just let me know.

GiedriusS replied:
Maybe they can still stay in this file? That doesn't really matter much IMO. My only complaint is that really all of these are methods of the OCI bucket struct so they naturally belong there.

GiedriusS reviewed (commented):
Almost there!
* The documentation needs updating (https://github.com/thanos-io/thanos/pull/4661/files#r798370792)
* Linter errors must be fixed
* Move "helper" functions to be struct methods (where struct methods are simply calling those functions)

aarontams commented:
> Almost there!
> 
> * The documentation needs updating (https://github.com/thanos-io/thanos/pull/4661/files#r798370792)
> * Linter errors must be fixed
> * Move "helper" functions to be struct methods (where struct methods are simply calling those functions)

Thanks @GiedriusS 
Will try to address your comments this weekend.

aarontams commented:
Hi @GiedriusS 
I think I have everything.

Also got a clean **golangci-lint run** with latest golangci-lint.
```
$ golangci-lint --version
golangci-lint has version 1.43.0 built from 861262b7 on 2021-11-03T11:57:46Z
```

Please let me know if I miss anything. Thx!

wiardvanrij commented:
Thanks again for your work. I think you need to run the `make check-docs` to fix formatting on the docs.

aarontams commented:
> Thanks again for your work. I think you need to run the `make check-docs` to fix formatting on the docs.

@wiardvanrij 
`make check-docs` complains about `https://about.sourcegraph.com/go/gophercon-2018-allocator-wrestling` link in storage.md which is not belong to my PR. It is belong to the existing `Reuse arrays` section.
Please advise
```
⣷ Formatting: /Users/aatam/go/src/github.com/aarontams/thanos/docs/storage.md...
error: 		docs/contributing/coding-style-guide.md:276: "https://about.sourcegraph.com/go/gophercon-2018-allocator-wrestling" not accessible; status code 400: Bad Request
```

aarontams commented:
Thanos end-to-end test failed. I don't think it is related to my PR.
Please let me know if this is not the case. Thx!
```=== CONT  TestCompactWithStoreGatewayWithPenaltyDedup
[2319](https://github.com/thanos-io/thanos/runs/5076757893?check_suite_focus=true#step:5:2319)
Error:     compact_test.go:838: compact_test.go:838:
[2320](https://github.com/thanos-io/thanos/runs/5076757893?check_suite_focus=true#step:5:2320)
        
[2321](https://github.com/thanos-io/thanos/runs/5076757893?check_suite_focus=true#step:5:2321)
         unexpected error: unable to find metrics [thanos_blocks_meta_synced] with expected values after 50 retries. Last error: <nil>. Last values: [43]
```

aarontams commented:
> Hi @GiedriusS I think I have everything.
> Also got a clean **golangci-lint run** with latest golangci-lint.
> ```
> $ golangci-lint --version
> golangci-lint has version 1.43.0 built from 861262b7 on 2021-11-03T11:57:46Z
> ```

@GiedriusS 
For some reason, the go-lint still failed according to the [report](https://github.com/thanos-io/thanos/runs/5076757841?check_suite_focus=true).
```
cannot 'detected files without copyright: you have unstaged changes.
[446](https://github.com/thanos-io/thanos/runs/5076757841?check_suite_focus=true#step:6:446)
M	pkg/objstore/oci/helper.go
[447](https://github.com/thanos-io/thanos/runs/5076757841?check_suite_focus=true#step:6:447)
M	pkg/objstore/oci/oci.go
[448](https://github.com/thanos-io/thanos/runs/5076757841?check_suite_focus=true#step:6:448)
Please commit or stash them.
```

I have copyright for both files. Please advise how to address it.
```
// Copyright (c) 2022, Oracle and/or its affiliates.
// Licensed under the Apache License 2.0.
```

BTW: I got clean result for `golangci-lint run`, but I couldn't get `make go-lint` running on my Mac.
```
$ make go-lint
all modules verified
>> verifying modules being imported
>> linting all of the Go files GOGC=
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0xb01dfacedebac1e pc=0x7fff6861770a]

runtime stack:
runtime: unexpected return pc for runtime.sigpanic called from 0x7fff6861770a
stack: frame={sp:0x7ffeefbfeec8, fp:0x7ffeefbfef18} stack=[0x7ffeefb7ff68,0x7ffeefbfefd0)
````

GiedriusS commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2022, Oracle and/or its affiliates.
+// Licensed under the Apache License 2.0.
```
This needs to be different: https://github.com/thanos-io/thanos/blob/main/scripts/copyright/copyright.go#L15-L21. When you are adding PRs to Thanos then you become a Thanos author (: You still retain the copyright which in this case is Oracle but just the header needs to be a bit different. I hope that's OK. Also, IANAL.

```suggestion
// Copyright (c) The Thanos Authors.
// Licensed under the Apache License 2.0.
```

aarontams replied:
Update.
Please let me know if there is anything else that need me to adjust. THX!

aarontams replied:
Hi @GiedriusS 
I have rebased my PR and updated the oci-go-sdk. Please let me know if the latest looks good to you.
Thx!

aarontams commented:
I am an Oracle employee. It is blocked by Oracle lawyer. I have been working with them for the past couple months. Hopefully, I will get the finally approval to contribute in the next couple months.

When it is clear by the Oracle Lawyer, I will rebase my PR.

aarontams commented on scripts/cfggen/main.go:
```diff
@@ -29,6 +29,7 @@ import (
 	"github.com/thanos-io/thanos/pkg/objstore/cos"
 	"github.com/thanos-io/thanos/pkg/objstore/filesystem"
 	"github.com/thanos-io/thanos/pkg/objstore/gcs"
+	"github.com/thanos-io/thanos/pkg/objstore/oci"
```
ignore

aarontams commented on scripts/cfggen/main.go:
```diff
@@ -29,6 +29,7 @@ import (
 	"github.com/thanos-io/thanos/pkg/objstore/cos"
 	"github.com/thanos-io/thanos/pkg/objstore/filesystem"
```
ignore

wiardvanrij commented on NOTICE:
```diff
@@ -0,0 +1,4 @@
+Thanos
+Copyright (c) 2022 The Thanos Authors
+
+This project includes software developed at Oracle America, Inc.
```
https://infra.apache.org/licensing-howto.html

> Do not modify LICENSE or NOTICE for non-bundled dependencies.

I rather see this go. If every person or company starts doing it, it becomes a mess.

aarontams replied:
Hi @wiardvanrij 
The guideline is from Oracle's legal counsel who is following the CFCF and Apache guideline.

CNCF doc: https://github.com/cncf/foundation/blob/main/copyright-notices.md 
Apache License  doc: https://infra.apache.org/licensing-howto.html

I will check with our legal team and see if there is anything that we can do on this.

Thx!

Djelibeybi replied:
Hey folks, I'm one of the open source advocates at Oracle and we are working with legal on this (and other) copyright-related requirements, but for now, this is a mandated requirement for our contributions to CNCF projects.

wiardvanrij replied:
_With my maintainer hat on;_ 

Sorry if I'm being hesitant but my priority is the project, community, and users. Perhaps it's not a big deal and this done soon.  As I said, I've pinged others as well to see what we have to do :) 

_With my personal hat on;_

Why? As noted in those guidelines you provide, it causes a burden on OSS projects/maintainers. Even your own employee is blocked for weeks/months with legal. It's IMO clear that you retain the copyright. 

That said, I honestly admire the effort here to contribute to the project. It's awesome to have quality code improvements from companies. Maybe we as a project have to do more to enable the "we got a legal department" companies in order to be more easily involved in open source without constraints. @Djelibeybi , I would personally appreciate the reasoning for the mandate. What is the legal thing that has to be covered here and what can we do as a project to make that easier? 

Anyhow, still thanks a lot @aarontams for your work (:  We will get this done!

Djelibeybi replied:
@wiardvanrij With (mostly) my personal hat on: there is no clear record of who "The Thanos Authors" _are_ outside of git that can be pointed at with zero ambiguity as to the provenance of the code _when it was contributed_ so when you say "it's .. clear that you retain the copyright" it's _not_ clear who "you" is in this case: it assumes that Git and GitHub are available for cross-reference.

But if this source code is distributed without any Git metadata, which happens a lot when packaged for distros, or when distributed as a container imagte, there is no way to prove the retention of copyright beyond "The Thanos Authors" and you're not distributing a handy list of those authors.

bwplotka replied:
Hi All, 

Unfortunately, such license entry is a precedence for us and might have unforeseen implications for our users. We are not lawyers, so we can't approve this. 

We added a ticket to CNCF to check if adding this line is possible, sorry for the inconvenience.

qedrakmar replied:
Hi all.  Just sticking my nose in here, so I can watch this one issue.  The CNCF's response here is particularly relevant to me.

wiardvanrij commented on docs/storage.md:
```diff
@@ -471,6 +472,72 @@ config:
+### OCI
+
+[Oracle Cloud Infrastructure](https://www.oracle.com/cloud). To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.
+You can get a free OCI Cloud Services by following the [Oracle Cloud Free Tier](https://www.oracle.com/cloud/free) page.
```
IMO, it's fine to link to docs like on line 477, 479, but this is more promotion / advertising. Let's keep this project neutral in that sense (i.e. we also don't add things about other cloud vendors free tiers).

aarontams replied:
Hi @wiardvanrij 
Understand. I just pushed another commit with updated description. Please take a look. Thanks for reviewing it.

wiardvanrij replied:
Thanks, added a new review. I will also test it soon btw, just haven't got the time for it yet.

Djelibeybi replied:
@wiardvanrij we have an OCI tenancy available for CNCF project testing purposes, so if you can point me towards your test framework, we can start looking into providing the resources for CI-triggered automatic testing on an ongoing basis too.

wiardvanrij reviewed (commented):
Thanks again for the work, I did have some (personal) feedback. I'll ask with others think of it as well, as I don't want to be the only one who has a thought about it.

wiardvanrij commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,270 @@
+	return objectNames, response.NextStartWith, nil
+}
+
+func (config *Config) validateConfig() (err error) {
```
nitpick; For usability, we can improve this by validating everything before returning. For example https://github.com/thanos-io/thanos/blob/main/pkg/objstore/azure/azure.go#L149

aarontams replied:
Good idea. Just pushed an updated.

wiardvanrij commented on docs/storage.md:
```diff
@@ -471,6 +472,68 @@ config:
+### OCI - Oracle Cloud Infrastructure
+
+To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage which supports three different providers, default, instance principal, and raw. 
+The default HTTP Client comes with this Thanos OCI Object Storage should be able to handle most of the use cases.
```
Let's try to make this more like a statement. Does it work or not? As a user, you don't want the "should" and "most of the use-cases" :D 

If it does work, we can rename this part towards something like "OCI has 3 different providers; x, y, and z, which are supported by this object storage client implementation" - Or something like that. 

It could also be I'm misreading this part, so feel free to correct me on that.

If there are any docs, I still think it's good to link towards that as well.

aarontams replied:
Make sense Just pushed an updated.

Djelibeybi commented:
@aarontams we have approval from Oracle legal to use the standard Thanos boilerplate copyright headers. Please go ahead and update your PR accordingly.

aarontams commented:
Thanks @Djelibeybi 
I will remove the NOTICE file and update the PR when I have a chance.

aarontams commented:
@bwplotka and @wiardvanrij 
Just did another update to my PR. I 
- removed the NOTICE file
- rebased to master
- ran TestObjStore_AcceptanceTest_e2e and passed

Please let me know if you guys have any more concerns or questions. Thx!

GiedriusS commented:
Seems like there is a conflict on `go.mod` @aarontams

aarontams commented:
> Seems like there is a conflict on `go.mod` @aarontams

Hi @GiedriusS 
My bad. Fixed my PR and rebased to master. Should work now.
Thanks for checking on that.

aarontams commented:
Fixed the go-lint issue. 
Should be good now.
Thx!

aarontams commented:
@GiedriusS, @bwplotka, and @wiardvanrij
Please let me know if you guys have any inputs to my PR.
Thx!

wiardvanrij commented:
Sorry for lag. I will review this soon. Meanwhile if you could fix the docs? 
```
error: 		docs/storage.md:49: link #oci, normalized to: link /home/runner/work/thanos/thanos/docs/storage.md#oci, existing ids: [object-storage--data-format configuring-access-to-object-storage how-to-use-our-special-config-flags supported-clients s3 s3-server-side-encryption credentials aws-policies sts-endpoint gcs using-googleapplicationcredentials using-inline-a-service-account gcs-policies azure openstack-swift tencent-cos aliyun-oss baidu-bos filesystem oci---oracle-cloud-infrastructure default-provider instance-principal-provider raw-provider how-to-add-a-new-client-to-thanos data-in-object-storage tsdb-block metadata-file-metajson external-labels index-format-index symbol-table series label-index postings label-offset-table postings-offset-table toc chunks-file-format chunk tombstones]: file exists, but does not have such id
```

Djelibeybi commented on docs/storage.md:
```diff
@@ -484,6 +485,68 @@ config:
+To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.
+OCI has 3 different providers; default, instance principal, and raw, which are supported by this object storage client implementation
+
+#### Default Provider
+The default config provider will look for configurations in 3 places: file in `$HOME/.oci/config`, `$HOME/.obmcs/config`, and
+variables names starting with the string TF_VAR. If the same configuration is found in multiple places the provider will
+prefer the first one.
```
```suggestion
#### API Signing Key

The default API signing key authentication provider leverages same [configuration as the OCI CLI](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/cliconcepts.htm) which is usually stored in at `$HOME/.oci/config` or via variable names starting with the string `OCI_CLI`. You can also use environment variables that start with `TF_VAR`. If the same configuration is found in multiple places the provider will prefer the first one.
```

aarontams replied:
Pushed the updates.

Djelibeybi commented on docs/storage.md:
```diff
@@ -484,6 +485,68 @@ config:
+variables names starting with the string TF_VAR. If the same configuration is found in multiple places the provider will
+prefer the first one.
+
+For Example:
```
```suggestion
The following example configures the provider to look for an existing API signing key for authentication:

```

aarontams replied:
Pushed the updates.

Djelibeybi commented on docs/storage.md:
````diff
@@ -484,6 +485,68 @@ config:
+  bucket: ""
+  compartment_ocid: ""
+```
+You can also include any of the optional configuration just like the example in `Default Provider`. 
+
+#### Raw Provider
+For Example:
+```yaml
+type: OCI
+config:
+  provider: "raw"
+  bucket: ""
+  compartment_ocid: ""
+  tenancy_ocid: ""
+  user_ocid: ""
+  region: ""
+  fingerprint: ""
+  privatekey: ""
+  passphrase: ""         // Optional passphrase to encrypt the private API Signing key
+```
+You can also include any of the optional configuration just like the example in `Default Provider`.
````
This block should be removed.

Djelibeybi commented on docs/storage.md:
````diff
@@ -484,6 +485,68 @@ config:
+    disable_compression: false    // Optional. If true, prevents the Transport from requesting compression.
+    client_timeout: 90s           // Optional time limit for requests made by the HTTP Client.
+```
+
+
````
```suggestion

Alternatively, you can provide API signing key authentication parameters in the configuration:

```yaml
type: OCI
config:
  provider: "raw"
  bucket: "thanos" // name of the object storage bucket
  compartment_ocid: "ocid1.compartment.oc1..."  // compartment in which the object storage bucket was created
  tenancy_ocid: "ocid1.tenancy.oc1..."
  user_ocid: "ocid1.user.oc1..."
  region: "us-ashburn-1"
  fingerprint: "aa:a1:bb:b1:cc:c1:dd:d1:ee:e1:ff:f1:aa:a2:bb:b2"
  privatekey: "/path/to/oci_api_key.pem"
  passphrase: ""         // Optional passphrase used to encrypt the private API Signing key

```

Djelibeybi commented on docs/storage.md:
````diff
@@ -484,6 +485,68 @@ config:
+```
+
+
+#### Instance Principal Provider
+For Example:
````
```suggestion
#### Instance Principal authentication

Instance principal authentication is available when the provider is running on a compute instance within Oracle Cloud Infrastructure.  Each compute instance has its own identity, and it authenticates using the certificates that are added to it. These certificates are automatically created, assigned to instances and rotated, preventing the need for you to distribute credentials to your hosts and rotate them. You are required to [configure an appropriate Dynamic Group and assign that group permissions via policies](https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm) so that the compute instance is authorised to access the object storage bucket.

The following configuration example enables instance principal authentication:
```

Djelibeybi commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,277 @@
+	var errMsg []string
+
+	if config.Tenancy == "" {
+		errMsg = append(errMsg, "no OCI tenancy ocid specified")
```
```suggestion
		errMsg = append(errMsg, "Tenancy OCID")
```

Djelibeybi commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,277 @@
+		errMsg = append(errMsg, "no OCI tenancy ocid specified")
+	}
+	if config.User == "" {
+		errMsg = append(errMsg, "no OCI user ocid specified")
```
```suggestion
		errMsg = append(errMsg, "User OCID")
```

Djelibeybi commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,277 @@
+		errMsg = append(errMsg, "no OCI user ocid specified")
+	}
+	if config.Region == "" {
+		errMsg = append(errMsg, "no OCI region specified")
```
```suggestion
		errMsg = append(errMsg, "region")
```

Djelibeybi commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,277 @@
+		errMsg = append(errMsg, "no OCI region specified")
+	}
+	if config.Fingerprint == "" {
+		errMsg = append(errMsg, "no OCI fingerprint specified")
```
```suggestion
		errMsg = append(errMsg, "API signing key fingerprint")
```

Djelibeybi commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,277 @@
+		errMsg = append(errMsg, "no OCI fingerprint specified")
+	}
+	if config.PrivateKey == "" {
+		errMsg = append(errMsg, "no OCI privatekey specified")
```
```suggestion
		errMsg = append(errMsg, "API signing key")
```

Djelibeybi commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	if len(errMsg) > 0 {
+		return errors.New(strings.Join(errMsg, ", "))
```
```suggestion
		return errors.New("Configuration parameters missing: " + strings.Join(errMsg, ", "))
```

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+	var err error
+
+	if err := yaml.Unmarshal(ociConfig, &config); err != nil {
+		return nil, errors.Wrapf(err, "unable to unmarshal the given oci configurations")
```
```suggestion
		return nil, errors.Wrapf(err, "unable to unmarshal the provided Oracle Cloud Infrastructure configuration")

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+	}
+
+	provider := Provider(strings.ToLower(config.Provider))
+	level.Info(logger).Log("msg", "creating OCI client", "provider", provider)
```
```suggestion
	level.Info(logger).Log("msg", "creating Oracle Cloud Infrastructure client", "provider", provider)
```

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+	case InstancePrincipalConfigProvider:
+		configurationProvider, err = auth.InstancePrincipalConfigurationProvider()
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to create OCI instance principal config provider")
```
```suggestion
			return nil, errors.Wrapf(err, "unable to create Oracle Cloud Infrastructure instance principal authentication provider")
```

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+		}
+	case RawConfigProvider:
+		if err := config.validateConfig(); err != nil {
+			return nil, errors.Wrapf(err, "invalid oci configurations")
```
```suggestion
			return nil, errors.Wrapf(err, "invalid Oracle Cloud Infrastructure configuration")
```

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+		configurationProvider = common.NewRawConfigurationProvider(config.Tenancy, config.User, config.Region,
+			config.Fingerprint, config.PrivateKey, &config.Passphrase)
+	default:
+		return nil, errors.Wrapf(err, fmt.Sprintf("unsupported OCI provider: %s", provider))
```
```suggestion
		return nil, errors.Wrapf(err, fmt.Sprintf("unsupported Oracle Cloud Infrastructure authentication provider: %s", provider))
```

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+
+	client, err := objectstorage.NewObjectStorageClientWithConfigurationProvider(configurationProvider)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to create ObjectStorage client with the given oci configurations")
```
```suggestion
		return nil, errors.Wrapf(err, "unable to create ObjectStorage client with the provided Oracle Cloud Infrastructure configuration")
```

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+	if err != nil {
+		return nil, err
+	}
+	level.Debug(logger).Log("msg", fmt.Sprintf("OCI tenancy namespace: %s", *namespace))
```
```suggestion
	level.Debug(logger).Log("msg", fmt.Sprintf("Oracle Cloud Infrastructure tenancy namespace: %s", *namespace))
```

aarontams replied:
Pushed the updates.

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+	ctx := context.Background()
+	bkt.name = objstore.CreateTemporaryTestBucketName(t)
+	if err := bkt.createBucket(ctx, config.Compartment); err != nil {
+		t.Errorf("failed to create temporary OCI bucket '%s' for OCI tests", bkt.name)
```
```suggestion
		t.Errorf("failed to create temporary Oracle Cloud Infrastructure bucket '%s' for testing", bkt.name)
```

aarontams replied:
Pushed the updates.

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+		return nil, nil, err
+	}
+
+	t.Logf("created temporary OCI bucket '%s' for OCI tests", bkt.name)
```
```suggestion
	t.Logf("created temporary Oracle Cloud Infrastructure bucket '%s' for testing", bkt.name)
```

aarontams replied:
Pushed the updates.

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+	return bkt, func() {
+		objstore.EmptyBucket(t, ctx, bkt)
+		if err := bkt.deleteBucket(ctx); err != nil {
+			t.Logf("failed to delete temporary OCI bucket %s for OCI tests: %s", bkt.name, err)
```
```suggestion
			t.Logf("failed to delete temporary Oracle Cloud Infrastructure bucket %s used for testing: %s", bkt.name, err)
```

aarontams replied:
Pushed the updates.

Djelibeybi commented on pkg/objstore/oci/oci.go:
```diff
@@ -0,0 +1,375 @@
+		if err := bkt.deleteBucket(ctx); err != nil {
+			t.Logf("failed to delete temporary OCI bucket %s for OCI tests: %s", bkt.name, err)
+		}
+		t.Logf("deleted temporary OCI bucket '%s' for OCI tests", bkt.name)
```
```suggestion
		t.Logf("deleted temporary Oracle Cloud Infrastructure bucket '%s' used for testing", bkt.name)
```

aarontams replied:
Pushed the updates.

Djelibeybi reviewed (changes requested):
Because the acronym OCI is overloaded, it would be better to expand out all the visible uses of OCI to Oracle Cloud Infrastructure to avoid any potential confusion. I've also had a go at tidying up the docs and error messages, but please test this before updating the PR! :)

aarontams commented on docs/storage.md:
````diff
@@ -484,6 +485,68 @@ config:
 prefix: ""
 ```
 
+### OCI - Oracle Cloud Infrastructure
````
Pushed the updates.

aarontams commented on docs/storage.md:
```diff
@@ -484,6 +485,68 @@ config:
+### OCI - Oracle Cloud Infrastructure
+
+To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.
+OCI has 3 different providers; default, instance principal, and raw, which are supported by this object storage client implementation
```
Pushed the updates.

wiardvanrij commented:
> @wiardvanrij FYI The doc check has 3 errors which are unrelated to my PR.
> 
> ```
> error: 3 errors:
> [125](https://github.com/thanos-io/thanos/runs/7124986832?check_suite_focus=true#step:5:126)
> 		docs/quick-tutorial.md:3: "https://katacoda.com/thanos/courses/thanos" not accessible; status code 404: Not Found
> [126](https://github.com/thanos-io/thanos/runs/7124986832?check_suite_focus=true#step:5:127)
> 		docs/getting-started.md:82: "https://katacoda.com/thanos/courses/thanos/1-globalview" not accessible; status code 404: Not Found
> [127](https://github.com/thanos-io/thanos/runs/7124986832?check_suite_focus=true#step:5:128)
> 		CONTRIBUTING.md:77: "https://katacoda.com/thanos" not accessible; status code 404: Not Found
> [128](https://github.com/thanos-io/thanos/runs/7124986832?check_suite_focus=true#step:5:129)
> ```

Yea, katacoda closed it's doors (: - We have fixed this on main now (https://github.com/thanos-io/thanos/pull/5454) , please have an other rebase (sorry about that :D )

wiardvanrij commented:
Also feel free to create a changelog entry already ^^

aarontams commented:
Hi @wiardvanrij 
Thanks again for fixing the doc in master.

Rebased. Hopefully, that will resolve the last failed doc check.

I also checked the `I added CHANGELOG entry for this change.` box.

aarontams commented:
FYI
The latest doc check error is not related to my PR.
```
error: 		docs/getting-started.md:114: "https://build.thebeat.co/monitoring-the-beat-microservices-a-tale-of-evolution-4e246882606e" not accessible even after retry; status code 0: Get "https://build.thebeat.co/monitoring-the-beat-microservices-a-tale-of-evolution-4e246882606e?gi=242b3805957d": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
[125](https://github.com/thanos-io/thanos/runs/7160297573?check_suite_focus=true#step:5:126)
make: *** [Makefile:243: check-docs] Error 1
```

GiedriusS commented on pkg/objstore/oci/helper.go:
```diff
@@ -0,0 +1,277 @@
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+func DefaultTransport(config Config) *http.Transport {
```
There's a common function for this now - could you please remove this and use that function?

aarontams replied:
Hi @GiedriusS,
Sorry for the late response, I am on vacation this week.

This `DefaultTransport` func allow user to define some customize setting here
```
type: OCI
config:
  provider: "default"
  bucket: ""
  compartment_ocid: ""
  part_size: ""                   // Optional part size to override the OCI default of 128 MiB, value is in bytes.
  max_request_retries: ""         // Optional maximum number of retries for a request.
  request_retry_interval: ""      // Optional sleep duration in seconds between retry requests.
  http_config:
    idle_conn_timeout: 1m30s      // Optional maximum amount of time an idle (keep-alive) connection will remain idle before closing itself. Zero means no limit.
    response_header_timeout: 2m   // Optional amount of time to wait for a server's response headers after fully writing the request.
    tls_handshake_timeout: 10s    // Optional maximum amount of time waiting to wait for a TLS handshake. Zero means no timeout.
    expect_continue_timeout: 1s   // Optional amount of time to wait for a server's first response headers. Zero means no timeout and causes the body to be sent immediately.
    insecure_skip_verify: false   // Optional. If true, crypto/tls accepts any certificate presented by the server and any host name in that certificate.
    max_idle_conns: 100           // Optional maximum number of idle (keep-alive) connections across all hosts. Zero means no limit.
    max_idle_conns_per_host: 100  // Optional maximum idle (keep-alive) connections to keep per-host. If zero, DefaultMaxIdleConnsPerHost=2 is used.
    max_conns_per_host: 0         // Optional maximum total number of connections per host.
    disable_compression: false    // Optional. If true, prevents the Transport from requesting compression.
    client_timeout: 90s           // Optional time limit for requests made by the HTTP Client.
```

Do you want me to `ONLY` use the `DefaultTransport` func in net/http/transport.go to construct the Transport for the httpclient?
If this is the case, I will need to remove the customize config fields (which are hard coded) from the PR.

Please advice.

kakkoyun replied:
> Sorry for the late response, I am on vacation this week.

Sorry @aarontams for the earlier message, just saw this one. Maybe you can open another PR against the new repo when you get back.

GiedriusS reviewed (dismissed):
There are some suggestions by @Djelibeybi to the documentation + one nit. Could you please look through those, rebase, and then we can merge this?

aarontams commented:
> There are some suggestions by @Djelibeybi to the documentation + one nit. Could you please look through those, rebase, and then we can merge this?

It was already done in previous update from 2 weeks ago Thx!

kakkoyun commented:
@aarontams Could you please rebase the PR? I would like to merge these changes and sync them to thanos-io/objstore before the #5254 is merged.

aarontams commented:
@kakkoyun @GiedriusS @wiardvanrij 
I am confused, I have rebased my PR in the last 24 hours. Is there any reason @kakkoyun needs to close my PR?

Djelibeybi commented:
@aarontams it was moved to https://github.com/thanos-io/objstore/issues/14

aarontams commented:
@Djelibeybi 
https://github.com/thanos-io/objstore/issues/14 is the issue in https://github.com/thanos-io/objstore. Not PR. My PR got closed here.

> @kakkoyun 
> @aarontams Could you please rebase the PR? I would like to merge these changes and sync them to thanos-io/objstore before the https://github.com/thanos-io/thanos/pull/5254 is merged.

I thought @kakkoyun wanted to merge this PR to https://github.com/thanos-io/thanos repo first. Instead this PR is closed. I am confused.

Djelibeybi commented:
Ah, right. Sorry, I just saw the link and didn't look too closely. Guess we'll have to wait and see.

yeya24 commented:
> I thought @kakkoyun wanted to merge this PR to https://github.com/thanos-io/thanos repo first. Instead this PR is closed. I am confused. Am I missing something obvious?

Sorry about that. I totally missed this pr and enabled auto-merge for https://github.com/thanos-io/thanos/pull/5510. This causes all the object-store related code to be removed from this codebase.

I think @kakkoyun intended to merge this pr first so that's why he asked you for a rebase. But I just merged that https://github.com/thanos-io/thanos/pull/5510 so that Kemal had to only close this pr. I am so sorry about it. @aarontams Would you mind opening a pr on the repo for this change? Thanks!

kakkoyun commented:
Sorry for not clearly communicating my actions here. It is exactly as @yeya24 explained. 

I saw that you're on vacation and I didn't want to push you to rebase. So I moved on. We still want to have these changes. That's why I have opened an issue on the new objstore repo for this.

I know it's a bit of more work. Could you please open a new PR on the new repo? I hadn't done it because the package structure has changed and it's more involved than syncing over some files.

Sorry for the inconvenience again @aarontams

aarontams commented:
Hi @kakkoyun and @yeya24 
No problem. With all the bumps that I have faced since I opened this PR last September, this little bump is not a big deal.

I have a new PR ready in https://github.com/thanos-io/objstore/pull/15 with new test result.

Hopefully, it will be a smooth ride from here. We can move the future conversation to the new [PR](https://github.com/thanos-io/objstore/pull/15).
